### PR TITLE
Feature/create offer interface

### PR DESCRIPTION
**Pull request by ahwei**
# Description

create offer interface
Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/utils/Secure.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import { getProgramBySeed } from '~/utils/signature'
+import { getStorage } from '~/utils/storage'
+
+import { retrieveSeed } from './extension'
+
+class Secure {
+    static getSeed = async () => {
+        const keyring = await getStorage('keyring')
+        if (!keyring) {
+            return
```
Comment (ahwei) [inline src/utils/Secure.ts]: throw error

File: src/utils/Secure.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        }
+        const password = await getDataFromMemory('password')
+        if (!password) {
+            return
```
Comment (ahwei) [inline src/utils/Secure.ts]: throw error

File: src/utils/Secure.ts (hunk 1 above)
Comment (hryjosn) [inline src/utils/Secure.ts]: Maybe can throw error

File: src/utils/Secure.ts (hunk 2 above)
Comment (hryjosn) [inline src/utils/Secure.ts]: Maybe can throw error

**Review by hryjosn — CHANGES_REQUESTED** [review]

File: src/utils/puzzles.ts [hunk 3]
```diff
@@ -2,13 +2,37 @@ import { Program } from '@rigidity/clvm'
...
...9ffff02ff26ffff04ff02ffff04ff0dffff04ff0bff808080808080ffff010b80ff0180ff0bff22ffff0bff2cff5880ffff0bff22ffff0bff22ffff0bff2cff5c80ff0580ffff0bff22ffff02ff32ffff04ff02ffff04ff07ffff04ffff0bff2cff2c80ff8080808080ffff0bff2cff8080808080ffff02ffff03ffff07ff0580ffff01ff0bffff0102ffff02ff2effff04ff02ffff04ff09ff80808080ffff02ff2effff04ff02ffff04ff0dff8080808080ffff01ff0bff2cff058080ff0180ffff04ffff04ff28ffff04ff5fff808080ffff02ff7effff04ff02ffff04ffff04ffff04ff2fff0580ffff04ff5fff82017f8080ffff04ffff02ff7affff04ff02ffff04ff0bffff04ff05ffff01ff808080808080ffff04ff17ffff04ff81bfffff04ff82017fffff04ffff0bff8204ffffff02ff36ffff04ff02ffff04ff09ffff04ff820affffff04ffff0bff2cff2d80ffff04ff15ff80808080808080ff8216ff80ffff04ff8205ffffff04ff820bffff808080808080808080808080ff02ff2affff04ff02ffff04ff5fffff04ff3bffff04ffff02ffff03ff17ffff01ff09ff2dffff0bff27ffff02ff36ffff04ff02ffff04ff29ffff04ff57ffff04ffff0bff2cff81b980ffff04ff59ff80808080808080ff81b78080ff8080ff0180ffff04ff17ffff04ff05ffff04ff8202ffffff04ffff04ffff04ff24ffff04ffff0bff7cff2fff82017f80ff808080ffff04ffff04ff30ffff04ffff0bff81bfffff0bff7cff15ffff10ff82017fffff11ff8202dfff2b80ff8202ff808080ff808080ff138080ff80808080808080808080ff018080'
```
Comment (hryjosn) [inline src/utils/puzzles.ts]: add constant to json file

**Review by hryjosn — CHANGES_REQUESTED** [review]

File: popup.html [hunk 4]
```diff
@@ -4,6 +4,7 @@
         <meta charset="utf-8" />
         <title>Hoogii</title>
         <script type="module" src="/src/popup/index.tsx"></script>
+        <meta name=”viewport”>
```
Comment (hryjosn) [inline popup.html]: wrong symbol

File: src/popup/pages/transaction.tsx [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            <div>
+                <div className="mb-3 text-left text-caption text-primary-100">
+                    Transaction
+                </div>
+                <div className="bg-box flex flex-col gap-1 px-2 py-3 shrink cursor-pointer rounded-sm ">
+                    <div className="text-left text-caption text-primary-100">
+                        Offer
+                    </div>
+                    {offerAssets.map((asset) => {
+                        const amount = !asset.assetId
```
Comment (hryjosn) [inline src/popup/pages/transaction.tsx]: reverse order of condition

File: src/utils/CAT.ts [hunk 6]
```diff
@@ -92,17 +93,20 @@ export class CAT extends Program {
         const change = sumSpendingValue - spendAmount
 
         const [firstCoin, ...restCoinList] = coinList
-
-        const puzzlehash = Program.fromBytes(
-            addressInfo(targetAddress).hash
-        ).toHex()
+        console.log(
```
Comment (hryjosn) [inline src/utils/CAT.ts]: remove console.log()

File: src/utils/Offer.ts [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+            return new CAT(
+                fromHex(sanitizeHex(assetId)),
+                puzzles.settlementPaymentsOld
+            )
+        } else {
+            return puzzles.settlementPaymentsOld
+        }
+    }
+
+    static async createCoinSpend() {}
```
Comment (hryjosn) [inline src/utils/Offer.ts]: remove this plz

File: src/utils/Offer.ts [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+                const {
+                    data: { data },
+                } = await callGetBalance({
+                    puzzle_hash: Program.fromBytes(cat.hash()).toHex(),
+                })
+
+                if (BigInt(data) < BigInt(offerPayment.amount)) {
+                    throw new Error("You don't have enough coin to spend")
+                }
+                console.log('settlement.hashHex()', settlement.hashHex())
```
Comment (hryjosn) [inline src/utils/Offer.ts]: remove console.log()

File: src/utils/Offer.ts [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        const spendList: CoinSpend[] = []
+        requestAssets.forEach(({ assetId, amount, memo }) => {
+            const coin: Coin = {
+                parent_coin_info:
+                    '0x0000000000000000000000000000000000000000000000000000000000000000',
+                puzzle_hash: this.generateSettlement(assetId).hashHex(),
+                amount: 0n,
+            }
+            const nonce = this.getNonce()
+            if (assetId) {
```
Comment (hryjosn) [inline src/utils/Offer.ts]: remove duplicated logic

File: src/utils/Offer.ts [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        }
+
+        const signatures = AugSchemeMPL.aggregate(
+            spendList
+                .filter((spend) => spend.coin.amount)
+                .map((item) =>
+                    Wallet.signCoinSpend(
+                        item,
+                        Buffer.from(
+                            'ae83525ba8d1dd3f09b277de18ca3e43fc0af20d20c4b3e92ef2a48bd291ccb2',
```
Comment (hryjosn) [inline src/utils/Offer.ts]: replace with constant: agg_sig_me_additional_data

File: src/utils/Secure.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        }
+        const seed = await retrieveSeed(password, keyring)
+
+        return seed
+    }
+
+    static getPuzzleReveal = async () => {
+        const seed = await this.getSeed()
+        if (seed) {
+            const puzzleReveal = getProgramBySeed(seed)
```
Comment (hryjosn) [inline src/utils/Secure.ts]: Rename variable as `puzzle`

File: src/utils/Wallet/types.ts [hunk 12]
```diff
@@ -9,7 +9,8 @@ export interface XCHPayload {
     amount: bigint
     memo?: string
     fee?: bigint
-    targetAddress: string
+    targetAddress?: string
```
Comment (hryjosn) [inline src/utils/Wallet/types.ts]: remove targetAddress

**Review by hryjosn — CHANGES_REQUESTED** [review]
Please modify request changes

File: src/api/extension/request.ts [hunk 13]
```diff
@@ -54,6 +54,10 @@ const accounts = async (): Promise<string[] | Errors.Error> => {
     return [account]
 }
 
+const createOffer = async (res: { offer: string }) => {
```
Comment (hryjosn) [inline src/api/extension/request.ts]: Continuely modify this

**Review by hryjosn — CHANGES_REQUESTED** [review]

**Review by hryjosn — APPROVED** [review]

**Pull request merged**